table/block_based/block_based_table_builder.cc

```diff
@@ -1886,10 +1899,76 @@ void BlockBasedTableBuilder::EnterUnbuffered() {
                                         r->pending_handle);
       }
     }
-
+    // TODO: figure out when we can ensure all the buffered data blocks are
+    // flushed and call this function
+    // TODO: figure out what is the right status message for successfully
+    // flushing all the buffered data blocks
+    if (ok()) {
+      FreeAllBufferedDataBlocksMemFromCache();
+    }
     std::swap(iter, next_block_iter);
   }
   r->data_block_buffers.clear();
+  r->buffer_exceeds_global_memory_limit = false;
+}
+
+// TODO figure out a better name for this function to reflect what it does in
+// the context of this class
+void BlockBasedTableBuilder::ReserveLastBufferedDataBlockMemInCache() {
+  // TODO-t figure out a better way to represent there is no need to reserve mem
+  // due to no cache
+  CacheRep* cache_rep = rep_->cache_rep;
+  if (cache_rep == nullptr) {
+    return;
+  }
+
+  // TODO figure out whether lock is needed for the cache
+
+  size_t kSizeDummyEntry = rep_->data_block_buffers.back().size();
+  Cache::Handle* handle = nullptr;
+  // TODO-l learn how cache key works
+  // TODO-l learn how (dummy) entries are deleted in cache
+  Status status = cache_rep->cache_->Insert(
+      cache_rep->GetNextCacheKey(), nullptr, kSizeDummyEntry,
+      GetNoopDeleterForRole<
+          CacheEntryRole::kCompressionDictionaryBuildingBuffer>(),
+      &handle);
+  // TODO figure out if we still need to push back a handle when insertion fails
```

Comment:
The API doc only says deleter needs to be called by client on failure, which is irrelevant to us since the value is nullptr. So I assume we don't need to do anything else like call `Release()`.